pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,282 @@
+	// exempt priority level would be the highest priority level. The field is default to false and only those system
+	// preset priority level can be exempt.
+	// +optional
+	Exempt *bool
+	// `assuredConcurrencyShares` is a positive number for a non-exempt priority level, representing the weight by which
+	// the priority level shares the concurrency from the global limit. The concurrency limit of an apiserver is divided
+	// among the non-exempt priority levels in proportion to their assured concurrency shares. Basically this produces
+	// the assured concurrency value (ACV) for each priority level:
+	//
+	//             ACV(l) = ceil( SCL * ACS(l) / ( sum[priority levels k] ACS(k) ) )
+	//
+	AssuredConcurrencyShares int32
+	// `queues` is a number of queues that belong to a non-exempt PriorityLevelConfiguration object. The queues exist
+	// independently at each apiserver. The value must be positive for a non-exempt priority level and setting it to 1
+	// disables shufflesharding and makes the distinguisher method irrelevant.
+	Queues int32
+	// `handSize` is a small positive number for applying shuffle sharding. When a request arrives at an apiserver the
+	// request flow identifier’s string pair is hashed and the hash value is used to shuffle the queue indices and deal
+	// a hand of the size specified here. If empty, the hand size will the be set to 1.
+	// +optional
+	HandSize *int32
+	// `queueLengthLimit` is a length limit applied to each queue belongs to the priority.  The value must be positive
+	// for a non-exempt priority level.
+	QueueLengthLimit int32
+}
+
+// PriorityLevelConfigurationConditionType is a valid value for PriorityLevelConfigurationStatusCondition.Type
+type PriorityLevelConfigurationConditionType string
+
+// PriorityLevelConfigurationStatus represents the current state of a "request-priority".
+type PriorityLevelConfigurationStatus struct {
+	// `conditions` is the current state of "request-priority".
+	Conditions []PriorityLevelConfigurationCondition
+}
+
+// PriorityLevelConfigurationCondition defines the condition of priority level.
+type PriorityLevelConfigurationCondition struct {
+	// `type` is the type of the condition.
+	Type PriorityLevelConfigurationConditionType
+	// `status` is the status of the condition.
+	// Can be True, False, Unknown.
+	Status corev1.ConditionStatus
+	// `lastTransitionTime` is the last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	// `reason` is a unique, one-word, CamelCase reason for the condition's last transition.
+	Reason string
+	// `message` is a human-readable message indicating details about last transition.
+	Message string
+}
+
+// Subject contains a reference to the object or user identities a role binding applies to.  This can either hold a direct API object reference,
+// or a value for non-objects such as user and group names.
```

Comment:
This comment is copied from rbac and is not quite right for our usage --- there are no role bindings in sight.

Also, I found this type in rbac confusing for a long time, despite the comment there.  I did not get the idea that the intent was to have a general object reference here, for the sake of easing future expansion, until I had a discussion on the mailling list.  I think it would be better to have a comment that more clearly emphasizes the idea.

Here is a suggested replacement comment:
```
// Subject matches a set of answers to the question "who/what is making this request?".
// Syntactically, Subject is a general API object reference.
// Authorization produces a username and a set of groups, and we imagine special kinds of non-namespaced objects,
// User and Group in API group "rbac.authorization.k8s.io", to represent them.
```

Is it best to use the API group "rbac.authorization.k8s.io" for _our_ imagined object kinds?  Or maybe we should just use rbac v1 Subject?

Reply:
Now the comment starts with "Subject contains a reference to the object or user identities".  This is still confusing.  In the authorization community at large we speak of "subject, verb, and object".  Saying that "subject" can refer to "the object" makes no sense in that light.  I think what this comment is trying to say is that "Subject" is either a real kube API object reference or a special sort of phony object reference that we define to refer to a username or user group name, but in all cases it is a reference to what the authorization community calls the "subject".  In https://kubernetes.io/docs/reference/access-authn-authz/rbac/#rolebinding-and-clusterrolebinding we have language that defines the corresponding concept as "a user or set of users".  So I suggest revising the comment.  Here is a suggested revision:

```
// Subject matches a set of users.
// Syntactically, Subject is a general API object reference.
// Authorization produces a username and a set of groups, and we imagine special kinds of non-namespaced objects,
// User and Group in API group "rbac.authorization.k8s.io", to represent such a username or group.
// The only kind of true object reference that currently will match any users is ServiceAccount.
```